### ISSUE TITLE
Running "spack load" right after "spack install" should work

### ISSUE THREAD

**Issue by HadrienG2** [spack/spack#10801]
Running "spack load <package>" right after "spack install <package>" will not work: the environment module system will not find the freshly installed package.

This can be worked around by re-sourcing the spack-env script, but I believe this is a usability papercut that spack load should be able to handle on its own.

### Steps to reproduce the issue

I most recently experienced this with templight, but every other package should behave similarly. So you may want to pick another that takes less time to build :smile: 

```console
$ spack install templight
$ spack load templight
Lmod Warning:  MODULEPATH is undefined.



Lmod has detected the following error:  The following module(s) are unknown: "templight-2019.01.09-gcc-8.2.1-hnbzriy"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore-cache load "templight-2019.01.09-gcc-8.2.1-hnbzriy"

Also make sure that all modulefiles written in TCL start with the string #%Module




$ source /opt/spack/share/spack/setup-env.sh
$ spack load templight
```

### Information on your system

```console
$ uname -a
Linux 688948042e36 4.9.0-8-amd64 #1 SMP Debian 4.9.144-3.1 (2019-02-19) x86_64 x86_64 x86_64 GNU/Linux
$ cat /etc/os-release 
NAME="openSUSE Tumbleweed"
# VERSION="20190301"
ID="opensuse-tumbleweed"
ID_LIKE="opensuse suse"
VERSION_ID="20190301"
PRETTY_NAME="openSUSE Tumbleweed"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:opensuse:tumbleweed:20190301"
BUG_REPORT_URL="https://bugs.opensuse.org"
HOME_URL="https://www.opensuse.org/"
$ zypper info lua-lmod
Loading repository data...
Reading installed packages...


Information for package lua-lmod:
---------------------------------
Repository     : openSUSE-Tumbleweed-Oss      
Name           : lua-lmod                     
Version        : 7.8.15-1.1                   
Arch           : x86_64                       
Vendor         : openSUSE                     
Installed Size : 1.1 MiB                      
Installed      : Yes                          
Status         : up-to-date                   
Source package : lua-lmod-7.8.15-1.1.src      
Summary        : Lua-based Environment Modules
Description    :                              
    Lmod is an Environment Module System based on Lua, Reads TCL Modules,
    Supports a Software Hierarchy.
```

**Comment by citibeth** [comment]
Have you run:
```
. spack/share/spack/setup-env.sh
```
as detailed at http://spack.io ?

As an aside, `spack load` is known to have two serious problems as way to construct desired environments in your `.bashrc` file:
 1. It is non-deterministic and can fail, even if it previously worked
 2. It is slow, and therefore not really suitable for use in `.bashrc` files.

`spack load` is good for one-offs trying things out.  Spack Environments are the preferred way to do this kind of thing reliably and quickly.  We should really update the docs...

**Comment by HadrienG2** [comment]
Indeed, I previously sourced setup-env in my `.profile`. If I didn't do so, the spack command would not be available.

My understanding is that whenever I source setup-env, it brings the module database in sync with all previously installed packages. I think this is something that spack load should be doing automatically, in order to avoid this unfortunate and beginner-hostile error.

> spack load is good for one-offs trying things out. Spack Environments are the preferred way to do this kind of thing reliably and quickly. We should really update the docs...

Please do so. At the moment, the best documentation available on using Spack-installed packages outside of Spack is https://spack.readthedocs.io/en/latest/workflows.html#running-binaries-from-packages , and it does not even mention Spack Environments. I cannot really use or recommend a feature whose only documentation resides in the CLI's --help switches.

**Comment by citibeth** [comment]
Hadrien,

Spack Environments are a relatively new feature, there is some
documentation, but we know we need better.  Our time is limited.  Would you
like to help?

-- Elizabeth

**Comment by HadrienG2** [comment]
I would be willing to help, but since I wasn't involved in this feature, I think I'm not the one best suited to write the first draft. I would effectively need to ask those who were involved in the development a bunch of questions, and they would spend more effort answering these questions than they would writing a draft of this documentation directly.

OTOH, once once a draft is around, I can probably do things like wording out bullet points into sentences, pointing out areas which are obscure to someone not already familiar with the feature...

**Comment by alalazo** [comment]
Fixed by @becker33 #14062 
```console
$ . share/spack/setup-env.sh
$ spack find
==> 0 installed packages
$ spack install zlib
==> Installing zlib
[ ... ]
==> Successfully installed zlib from binary cache
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu18.04-broadwell/gcc-9.0.1/zlib-1.2.11-os3djb57dqdzhaqsl2n7iefz3gwombyi
$ spack load zlib
$ echo $LD_LIBRARY_PATH
/home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu18.04-broadwell/gcc-9.0.1/zlib-1.2.11-os3djb57dqdzhaqsl2n7iefz3gwombyi/lib
```
